Allow batch size, period and queue limit to be set

benfoster opened the pull request:
This PR extends the sink to allow the batch size, period and queue limit to be set. Previously these were hardcoded values which we found unsuitable for a number of applications generating a large number of log entries. 

I'm still not 100% sure how configuration (e.g. appsettings.json) is auto-magically bound to the sink so have just followed the existing conventions.

Note that I had to use two constructors since the periodic batching only allows a queue limit to either be set or not (resulting in an unbound queue).

There's a couple of unreferenced JSON related fields which I look like they could be removed but again I'm unaware of they're being used for any dynamic binding.

ogaca-dd commented:
Hi @benfoster ,

Thank you for submitting this pull request. I will review it at the beginning of next week.

ogaca-dd commented on src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogSink.cs:
```diff
@@ -50,6 +51,27 @@ public DatadogSink(string apiKey, string source, string service, string host, st
+        public DatadogSink(string apiKey, string source, string service, string host, string[] tags, DatadogConfiguration config, int queueLimit, int? batchSizeLimit = null, TimeSpan? batchPeriod = null)
+            : base(batchSizeLimit ?? DefaultBatchSizeLimit, batchPeriod ?? DefaultBatchPeriod, queueLimit)
+        {
+            if (config.UseTCP)
```
Nice to have: Factorize the common code in both constructors into a single method (From line 57 to line 64)

benfoster replied:
Done

ogaca-dd commented:
> I'm still not 100% sure how configuration (e.g. appsettings.json) is auto-magically bound to the sink so have just followed the existing conventions.

I have just checked it and it looks good.

benfoster commented on src/Serilog.Sinks.Datadog.Logs/Configuration/Extensions/Microsoft.Extensions.Configuration/LoggerConfigurationDatadogLogsExtensions.cs:
```diff
@@ -50,8 +60,9 @@ public static class LoggerConfigurationDatadogExtensions
+            var sink = DatadogSink.Create(apiKey, source, service, host, tags, config, batchSizeLimit, batchPeriod, queueLimit);
 
-            return loggerConfiguration.Sink(new DatadogSink(apiKey, source, service, host, tags, config), logLevel);
+            return loggerConfiguration.Sink(sink);
```
Actually no, I removed this by mistake. I've now fixed.

benfoster commented on src/Serilog.Sinks.Datadog.Logs/Configuration/Extensions/System.Configuration/LoggerConfigurationDatadogLogsExtensions.cs:
```diff
@@ -47,7 +57,9 @@ public static class LoggerConfigurationDatadogExtensions
-            return loggerConfiguration.Sink(new DatadogSink(apiKey, source, service, host, tags, configuration), logLevel);
+            var sink = DatadogSink.Create(apiKey, source, service, host, tags, configuration, batchSizeLimit, batchPeriod, queueLimit);
+
+            return loggerConfiguration.Sink(sink);
```
Fixed

benfoster commented:
@ogaca-dd thanks for the review. I've addressed the comments on the PR.

I've reinstated the minimum log level with the slight change that I've renamed the property as I'm not sure this was actually working as intended before as we found setting `restrictedToMinimumLevel` in appsettings.json had no effect (most likely due to the property name binding).

ben-foster-cko commented on src/Serilog.Sinks.Datadog.Logs/Serilog.Sinks.Datadog.Logs.csproj:
```diff
@@ -1,7 +1,7 @@
 
   <PropertyGroup>
-    <TargetFrameworks>netstandard2.0;netstandard1.3;net45</TargetFrameworks>
+    <TargetFrameworks>netstandard2.0;netstandard1.3;</TargetFrameworks>
```
Reverted. Pulled this out so it would build on my machine (osx)

ben-foster-cko commented on src/Serilog.Sinks.Datadog.Logs/Configuration/Extensions/Microsoft.Extensions.Configuration/LoggerConfigurationDatadogLogsExtensions.cs:
```diff
@@ -45,7 +45,7 @@ public static class LoggerConfigurationDatadogExtensions
             DatadogConfiguration configuration = null,
             IConfigurationSection configurationSection = null,
-            LogEventLevel logLevel = LevelAlias.Minimum,
+            LogEventLevel restrictedToMinimumLevel = LevelAlias.Minimum,
```
We had previously observed that the sink log level settings wasn't working for us and I made an assumption based on other sinks that use the `restrictedToMinimumLevel` setting that the field must be named the same. I've reverted the change but please can you confirm that the log level works for you (using string based value e.g. `Debug`)?

ben-foster-cko replied:
Would also be good to know how you are testing this locally?

ogaca-dd replied:
I tested locally with the following code and `appsettings.json`
```C#
            var configuration = new ConfigurationBuilder()
                    .SetBasePath(Directory.GetCurrentDirectory())
                    .AddJsonFile("appsettings.json")
                    .Build();

            var seriLoglogger = new LoggerConfiguration()
                .ReadFrom.Configuration(configuration)
                .CreateLogger();

            seriLoglogger.Information("Test Information");
            seriLoglogger.Debug("Test Debug");
            System.Threading.Thread.Sleep(3000); // Give enought time to flush
```
```
// appsettings.json
{
    "Serilog": {
        "Using": [
            "Serilog.Sinks.Console",
            "Serilog.Sinks.Datadog.Logs"
        ],
        "MinimumLevel": "Debug",
        "WriteTo": [
            {
                "Name": "Console"
            },
            {
                "Name": "DatadogLogs",
                "Args": {
                    "apiKey": "<API_KEY>",
                    "source": "SerilogTest",
                    "host": "MyHost",
                    "logLevel": 2,
                    "batchSizeLimit": 42,
                    "queueLimit": 51,
                    "batchPeriod": null
                }
            }
        ],
        "Enrich": [
            "FromLogContext",
            "WithMachineName",
            "WithThreadId"
        ],
        "Properties": {
            "Application": "Sample"
        }
    }
}
```

When I set the log level to `Information` `"logLevel": 2`,  only the log `Test Information` is available in the DataDog Web App.
When I set the log level to `Debug` `"logLevel": 1`, both `Test Information` and `Test Debug` are available in DataDog Web App.

ben-foster-cko replied:
Do the string based values work instead (which is more common with Serilog sinks)?

IrisClasson commented:
Are there rate limits that we have to keep in mind when changing the periodic timer to a lower value, or if we were to use an IEventLogSink instead of the periodic batching? The API has many restrictions/limits but it’s unclear if they apply to sending logs entries (https://docs.datadoghq.com/api/ ). Requests with rate limits return rate limit information in the header

ben-foster-cko commented:
@ogaca-dd you're most welcome. Any ideas why this may be merged and released?

ogaca-dd commented:
@IrisClasson: As far as I know, there is no rate limits for sending logs. Sending logs are not free of charge.

ogaca-dd commented:
@ben-foster-cko : [Serilog.Sinks.Datadog.Logs 0.3.2](https://www.nuget.org/packages/Serilog.Sinks.Datadog.Logs/0.3.2) includes this PR.
